src/self_driving/planning/pilot_util.cpp

```diff
@@ -419,4 +425,233 @@ void PilotUtil::GroupFeaturesByOU(
   }
 }
 
+void PilotUtil::ComputeTableSizeRatios(const WorkloadForecast *forecast,
+                                       common::ManagedPointer<task::TaskManager> task_manager,
+                                       common::ManagedPointer<transaction::TransactionManager> txn_manager,
+                                       common::ManagedPointer<catalog::Catalog> catalog,
+                                       pilot::MemoryInfo *memory_info) {
+  NOISEPAGE_ASSERT(task_manager, "ComputeTableSizeRatios() requires task manager");
+
+  // Maps from table oid to the number of rows (acquired from pg_statistic)
+  std::unordered_map<catalog::table_oid_t, uint64_t> table_sizes;
+
+  auto to_row_fn = [&table_sizes, memory_info](const std::vector<execution::sql::Val *> &values) {
+    auto table_oid = static_cast<execution::sql::Integer *>(values[0])->val_;
+    auto num_rows = static_cast<execution::sql::Integer *>(values[1])->val_;
+
+    table_sizes[catalog::table_oid_t(table_oid)] = num_rows;
+    // Currently only track tables with more than 0 rows in the stats
+    memory_info->table_oids_.emplace_back(table_oid);
+  };
+
+  // Get <table_id, num_rows> pairs with num_rows > 0
+  auto query = fmt::format(
+      "select starelid, max(stanumrows) as num_rows from pg_statistic group by starelid "
+      "having max(stanumrows) > 0;");
+
+  common::Future<task::DummyResult> sync;
+  task_manager->AddTask(std::make_unique<task::TaskDML>(catalog::INVALID_DATABASE_OID, query,
+                                                        std::make_unique<optimizer::TrivialCostModel>(), false,
+                                                        to_row_fn, common::ManagedPointer(&sync)));
+
+  auto future_result = sync.WaitFor(Pilot::FUTURE_TIMEOUT);
+  if (!future_result.has_value()) {
+    throw PILOT_EXCEPTION("Future timed out.", common::ErrorCode::ERRCODE_IO_ERROR);
+  }
+  UNUSED_ATTRIBUTE bool success = future_result->second;
+  NOISEPAGE_ASSERT(success, "Failed reading from pg_statistic");
+
+  // Calculating queries' changes in the number of rows
+  auto &metadata = forecast->GetWorkloadMetadata();
+  auto txn = txn_manager->BeginTransaction();
+  // <query_id, <table_id, number of rows changed by that query in that table>>
+  std::unordered_map<execution::query_id_t, std::pair<catalog::table_oid_t, int64_t>> query_row_changes;
+  for (const auto &[query_id, query_text] : metadata.query_id_to_text_) {
+    // Parse the query
+    auto parse_result = parser::PostgresParser::BuildParseTree(query_text);
+    if (parse_result->NumStatements() == 0) continue;
+    auto statement = parse_result->GetStatement(0);
+
+    // Identify the table and the number of inserts/deletes
+    int64_t num_row_delta = 0;
+    catalog::table_oid_t table_oid;
+    auto accessor = catalog->GetAccessor(common::ManagedPointer(txn),
+                                         catalog::db_oid_t(forecast->GetDboidByQid(query_id)), DISABLED);
```

Comment:
I think this "works" for read-only transactions, but I don't know if there are consequences we need to be concerned about doing cross-database reads inside of a single transaction.  @lmwnshn or @mbutrovich?

Also, getting the accessor is probably expensive enough that doing this inside the loop for every query could be painful on performance.